File: lib/data/datasources/local/notification_service.dart
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:timezone/timezone.dart' as tz;
import 'package:timezone/data/latest.dart' as tz_data;
import '../../../domain/entities/habit.dart';
import '../../../domain/utils/seeded_rng.dart';

class NotificationService {
  static final NotificationService shared = NotificationService._();
  NotificationService._();

  final _plugin = FlutterLocalNotificationsPlugin();
  bool isAuthorized = false;

  static const _reminderMessages = [
    'Your tribute is waiting. Just a moment with God today.',
    'A few minutes. A small gift. God sees it all.',
    "Today's offering is ready whenever you are.",
    'Even a single check-in changes the shape of your day.',
    'Your habits are waiting — each one a gift to God.',
    'A quiet moment with God today. That\'s all it takes.',
    'Start with gratitude. Everything else follows.',
    'God meets you in the effort and in the rest.',
    'One small step today. He\'s already walking with you.',
    'Your tribute matters — even on the hard days.',
  ];

  static const _timeMilestones = [
    (7, 'One week down', "You showed up, and God met you in it. Let's keep going."),
    (14, 'Two weeks', "You're past the point where most people quit a new app. You're still here."),
    (21, '3 weeks in', "This is usually when the newness fades and it starts to feel like work. That's normal. You're doing the hard part."),
    (30, 'A full month', "30 days of giving this to God. That's not a small thing."),
    (45, 'Over 6 weeks', 'Most people never make it this far. The rhythm is starting to take hold. Keep showing up.'),
    (66, '66 days', "Research says this is roughly when a habit becomes automatic. This isn't something you do anymore — it's who you are."),
    (100, '100 days', "A hundred times you chose to show up. That's a life that's being changed."),
    (365, 'A full year', "365 days of giving this to God. Whatever this year looked like — the strong weeks and the quiet ones — He was in all of it."),
  ];

  Future<void> init() async {
    tz_data.initializeTimeZones();
    const androidSettings = AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: false,
      requestBadgePermission: false,
      requestSoundPermission: false,
    );
    await _plugin.initialize(const InitializationSettings(android: androidSettings, iOS: iosSettings));
    await checkAuthorization();
  }

  Future<void> checkAuthorization() async {
    final androidPlugin = _plugin.resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>();
    if (androidPlugin != null) {
      final granted = await androidPlugin.areNotificationsEnabled() ?? false;
      isAuthorized = granted;
    } else {
      // iOS — assume authorized if we've been granted before
      final prefs = await SharedPreferences.getInstance();
      isAuthorized = prefs.getBool('tribute_notifications_authorized') ?? false;
    }
  }

  Future<bool> requestAuthorization() async {
    final iosPlugin = _plugin.resolvePlatformSpecificImplementation<IOSFlutterLocalNotificationsPlugin>();
    if (iosPlugin != null) {
      final granted = await iosPlugin.requestPermissions(alert: true, badge: true, sound: true) ?? false;
      isAuthorized = granted;
      final prefs = await SharedPreferences.getInstance();
      await prefs.setBool('tribute_notifications_authorized', granted);
      return granted;
    }
    final androidPlugin = _plugin.resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>();
    if (androidPlugin != null) {
      final granted = await androidPlugin.requestNotificationsPermission() ?? false;
      isAuthorized = granted;
      return granted;
    }
    return false;
  }

  Future<void> clearBadge() async {
    await _plugin.cancel(0);
  }

  Future<void> scheduleDailyReminders() async {
    final prefs = await SharedPreferences.getInstance();
    final enabled = prefs.getBool('tribute_reminders_enabled') ?? false;
    if (!enabled) { await cancelDailyReminders(); return; }

    final hour = prefs.getInt('tribute_reminder_hour') ?? 8;
    final minute = prefs.getInt('tribute_reminder_minute') ?? 0;

    await cancelDailyReminders();

    for (int i = 0; i < 7; i++) {
      final body = _reminderMessages[i % _reminderMessages.length];
      await _plugin.zonedSchedule(
        100 + i,
        'Tribute',
        body,
        _nextWeekday(i + 1, hour, minute),
        NotificationDetails(
          android: const AndroidNotificationDetails('daily_reminder', 'Daily Reminders', importance: Importance.defaultImportance),
          iOS: const DarwinNotificationDetails(),
        ),
        androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
        uiLocalNotificationDateInterpretation: UILocalNotificationDateInterpretation.absoluteTime,
        matchDateTimeComponents: DateTimeComponents.dayOfWeekAndTime,
      );
    }
  }

  Future<void> cancelDailyReminders() async {
    for (int i = 0; i < 7; i++) {
      await _plugin.cancel(100 + i);
    }
  }

  Future<void> refreshAllNotifications(List<Habit> habits) async {
    await checkAuthorization();
    if (!isAuthorized) return;
    await scheduleDailyReminders();

    final prefs = await SharedPreferences.getInstance();
    final onboardingMs = prefs.getInt('tribute_onboarding_date');
    final onboarding = onboardingMs != null ? DateTime.fromMillisecondsSinceEpoch(onboardingMs) : DateTime.now();
    final daysSince = DateTime.now().difference(onboarding).inDays;

    for (final habit in habits) {
      await _scheduleTimeMilestones(habit);
    }
    await _scheduleVariableReinforcement(daysSince);
  }

  Future<void> _scheduleTimeMilestones(Habit habit) async {
    final prefs = await SharedPreferences.getInstance();
    final hour = prefs.getInt('tribute_reminder_hour') ?? 9;
    final minute = prefs.getInt('tribute_reminder_minute') ?? 0;

    for (final m in _timeMilestones) {
      final target = habit.createdAt.add(Duration(days: m.$1));
      if (!target.isAfter(DateTime.now())) continue;
      final scheduled = DateTime(target.year, target.month, target.day, hour, minute);
      await _plugin.zonedSchedule(
        Object.hash(habit.id, m.$1) & 0x7fffffff,
        m.$2, m.$3,
        _toTZDateTime(scheduled),
        NotificationDetails(
          android: const AndroidNotificationDetails('milestones', 'Milestones', importance: Importance.high),
          iOS: const DarwinNotificationDetails(),
        ),
        androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
        uiLocalNotificationDateInterpretation: UILocalNotificationDateInterpretation.absoluteTime,
      );
    }
  }

  Future<void> _scheduleVariableReinforcement(int daysSinceOnboarding) async {
    if (daysSinceOnboarding < 30) return;
    await _plugin.cancel(99999);

    const messages = [
      'Remember when this felt hard? Look at you now.',
      "You're building something real. God sees every single day.",
      "Your consistency is its own kind of worship. Keep going.",
      "The rhythm you've built? That's not willpower — that's faithfulness.",
      "Some days it's easy, some days it's not. You show up either way. That matters.",
    ];

    // Use a deterministic seeded RNG keyed on daysSinceOnboarding so the
    // same run produces the same schedule, avoiding message/interval drift
    // across multiple calls within the same app session.
    final rng = SeededRng(daysSinceOnboarding * 13 + 7);
    final daysUntilNext = 14 + (rng.next() % 15);
    final messageIdx = rng.next() % messages.length;
    final target = DateTime.now().add(Duration(days: daysUntilNext));
    final prefs = await SharedPreferences.getInstance();
    final hour = prefs.getInt('tribute_reminder_hour') ?? 10;
    final minute = prefs.getInt('tribute_reminder_minute') ?? 0;
    final scheduled = DateTime(target.year, target.month, target.day, hour, minute);
    final body = messages[messageIdx];

    await _plugin.zonedSchedule(
      99999, 'Tribute', body,
      _toTZDateTime(scheduled),
      NotificationDetails(
        android: const AndroidNotificationDetails('encouragement', 'Encouragement', importance: Importance.defaultImportance),
        iOS: const DarwinNotificationDetails(),
      ),
      androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
      uiLocalNotificationDateInterpretation: UILocalNotificationDateInterpretation.absoluteTime,
    );
  }

  // Helpers — using local timezone via DateTime
  // In production, integrate timezone package for precise scheduling

  tz.TZDateTime _toTZDateTime(DateTime dt) {
    return tz.TZDateTime.from(dt, tz.local);
  }

  tz.TZDateTime _nextWeekday(int weekday, int hour, int minute) {
    var now = tz.TZDateTime.now(tz.local);
    var scheduled = tz.TZDateTime(tz.local, now.year, now.month, now.day, hour, minute);
    while (scheduled.weekday != weekday || scheduled.isBefore(now)) {
      scheduled = scheduled.add(const Duration(days: 1));
    }
    return scheduled;
  }
}

File: swift_reference/Services/NotificationService.swift
import Foundation
import UserNotifications

@Observable
@MainActor
class NotificationService {
    static let shared = NotificationService()

    var isAuthorized: Bool = false

    private let center = UNUserNotificationCenter.current()
    private let dailyReminderCategory = "DAILY_REMINDER"
    private let milestoneCategory = "MILESTONE"

    private let reminderMessages: [String] = [
        "Your tribute is waiting. Just a moment with God today.",
        "A few minutes. A small gift. God sees it all.",
        "Today's offering is ready whenever you are.",
        "Even a single check-in changes the shape of your day.",
        "Your habits are waiting — each one a gift to God.",
        "A quiet moment with God today. That's all it takes.",
        "Start with gratitude. Everything else follows.",
        "God meets you in the effort and in the rest.",
        "One small step today. He's already walking with you.",
        "Your tribute matters — even on the hard days.",
    ]

    private let timeMilestoneMessages: [(day: Int, title: String, body: String)] = [
        (7, "One week down", "You showed up, and God met you in it. Let's keep going."),
        (14, "Two weeks", "You're past the point where most people quit a new app. You're still here."),
        (21, "3 weeks in", "This is usually when the newness fades and it starts to feel like work. That's normal. You're doing the hard part."),
        (30, "A full month", "30 days of giving this to God. That's not a small thing."),
        (45, "Over 6 weeks", "Most people never make it this far. The rhythm is starting to take hold. Keep showing up."),
        (66, "66 days", "Research says this is roughly when a habit becomes automatic. This isn't something you do anymore — it's who you are."),
        (100, "100 days", "A hundred times you chose to show up. That's a life that's being changed."),
        (365, "A full year", "365 days of giving this to God. Whatever this year looked like — the strong weeks and the quiet ones — He was in all of it."),
    ]

    func checkAuthorization() async {
        let settings = await center.notificationSettings()
        isAuthorized = settings.authorizationStatus == .authorized
    }

    func requestAuthorization() async -> Bool {
        do {
            let granted = try await center.requestAuthorization(options: [.alert, .badge, .sound])
            isAuthorized = granted
            return granted
        } catch {
            return false
        }
    }

    func clearBadge() {
        center.setBadgeCount(0)
    }

    func scheduleDailyReminders() {
        let enabled = UserDefaults.standard.bool(forKey: "tribute_reminders_enabled")
        guard enabled else {
            cancelDailyReminders()
            return
        }

        let hour = UserDefaults.standard.integer(forKey: "tribute_reminder_hour")
        let minute = UserDefaults.standard.integer(forKey: "tribute_reminder_minute")
        let reminderHour = (hour == 0 && minute == 0 && !UserDefaults.standard.bool(forKey: "tribute_reminders_enabled")) ? 8 : hour

        cancelDailyReminders()

        for i in 0..<7 {
            let content = UNMutableNotificationContent()
            content.title = "Tribute"
            content.body = reminderMessages[i % reminderMessages.count]
            content.sound = .default
            content.categoryIdentifier = dailyReminderCategory
            content.badge = 1

            var dateComponents = DateComponents()
            dateComponents.hour = reminderHour
            dateComponents.minute = minute
            dateComponents.weekday = i + 1

            let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: true)
            let request = UNNotificationRequest(
                identifier: "daily_reminder_\(i)",
                content: content,
                trigger: trigger
            )

            center.add(request)
        }
    }

    func cancelDailyReminders() {
        let ids = (0..<7).map { "daily_reminder_\($0)" }
        center.removePendingNotificationRequests(withIdentifiers: ids)
    }

    func scheduleTimeMilestoneNotifications(habitCreatedAt: Date, habitName: String) {
        let calendar = Calendar.current
        let hour = UserDefaults.standard.integer(forKey: "tribute_reminder_hour")
        let minute = UserDefaults.standard.integer(forKey: "tribute_reminder_minute")
        let notifHour = (hour == 0 && minute == 0) ? 9 : hour
        let notifMinute = (hour == 0 && minute == 0) ? 0 : minute

        for milestone in timeMilestoneMessages {
            guard let targetDate = calendar.date(byAdding: .day, value: milestone.day, to: calendar.startOfDay(for: habitCreatedAt)) else { continue }

            if targetDate <= Date() { continue }

            var components = calendar.dateComponents([.year, .month, .day], from: targetDate)
            components.hour = notifHour
            components.minute = notifMinute

            let content = UNMutableNotificationContent()
            content.title = milestone.title
            content.body = milestone.body
            content.sound = .default
            content.categoryIdentifier = milestoneCategory

            let trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
            let identifier = "milestone_\(habitName.lowercased().replacingOccurrences(of: " ", with: "_"))_day\(milestone.day)"
            let request = UNNotificationRequest(
                identifier: identifier,
                content: content,
                trigger: trigger
            )

            center.add(request)
        }
    }

    func scheduleVariableReinforcement(daysSinceOnboarding: Int) {
        guard daysSinceOnboarding >= 30 else { return }

        center.removePendingNotificationRequests(withIdentifiers: ["variable_reinforcement"])

        let daysUntilNext = Int.random(in: 14...28)
        let calendar = Calendar.current
        guard let targetDate = calendar.date(byAdding: .day, value: daysUntilNext, to: Date()) else { return }

        let messages = [
            "Remember when this felt hard? Look at you now.",
            "You're building something real. God sees every single day.",
            "Your consistency is its own kind of worship. Keep going.",
            "The rhythm you've built? That's not willpower — that's faithfulness.",
            "Some days it's easy, some days it's not. You show up either way. That matters.",
        ]

        let content = UNMutableNotificationContent()
        content.title = "Tribute"
        content.body = messages.randomElement() ?? messages[0]
        content.sound = .default

        let hour = UserDefaults.standard.integer(forKey: "tribute_reminder_hour")
        let minute = UserDefaults.standard.integer(forKey: "tribute_reminder_minute")
        var components = calendar.dateComponents([.year, .month, .day], from: targetDate)
        components.hour = hour == 0 ? 10 : hour
        components.minute = minute

        let trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
        let request = UNNotificationRequest(
            identifier: "variable_reinforcement",
            content: content,
            trigger: trigger
        )

        center.add(request)
    }

    func refreshAllNotifications(habits: [Habit]) async {
        await checkAuthorization()
        guard isAuthorized else { return }

        scheduleDailyReminders()

        let onboardingDate = UserDefaults.standard.object(forKey: "tribute_onboarding_date") as? Date ?? Date()
        let daysSince = Calendar.current.dateComponents([.day], from: onboardingDate, to: Date()).day ?? 0

        for habit in habits {
            scheduleTimeMilestoneNotifications(habitCreatedAt: habit.createdAt, habitName: habit.name)
        }

        scheduleVariableReinforcement(daysSinceOnboarding: daysSince)
    }
}
